### ISSUE TITLE
Globals defined in separate files are reported as undefined_variable

### ISSUE THREAD

**Issue by SvenMarcus** [Kampfkarren/selene#214]
### Description:
Selene reports globals defined in separate files as undefined variables when importing them into other files

### Example

File: lib.lua
```lua
function my_func(str)
    print(str)
end
```

File: test.lua
```lua
require("lib")

my_func("bla")
```

Result:
```
error[undefined_variable]: `my_func` is not defined
  ┌─ test.lua:3:1 
  │
3 │ my_func("bla")
  │ ^^^^^^^     
```

**Comment by Kampfkarren** [comment]
This is a bug for sure, but this one is really tricky to solve without some concerns--selene doesn't follow requires at all right now, making it do so would likely either increase complexity of the code by making it cache every file it comes across, or just grabbing it again which would be bad on performance. Plus, this doesn't even mention all the things require could actually be referring to.

**Comment by SvenMarcus** [comment]
If I remember correctly luacheck doesn't follow requires either, but rather assumes the global variable is valid if it was set in one of the files it was given to analyze.

That approach would also fit the goal of not finding all issues while making sure to only report valid ones

**Comment by Kampfkarren** [comment]
Can you detail that more? I think you're saying to just assume any variable after a require is defined, which would be in the category of never lying, but also feels like it would make the undefined_variable lint pretty useless if Selene just assumes that.

**Comment by SvenMarcus** [comment]
No, not any variable. I was thinking of building a cache of global variables that are defined in the files that selene analyses.
So in my example in the first post I would pass the folder that contains both Lua files to selene. It would then analyse the files and store the global `my_func` function in the cache for global variables. In test.lua selene could just assume the global variable is valid, because it's cached. 
It's probably best to ignore the require for that, since it's pretty hard to know at what point in a larger program a global variable becomes valid. If you're lucky it was defined at the top level of some file, but there could just as well be a function that sets it. And who knows when or if that's getting called.

So to summarize: cache all globals from the analyzed files, defined at the top level or in functions, and assume all globals that are in the cache are valid at any point.

**Comment by Kampfkarren** [comment]
Ah, yeah that's more or less what I meant by "making it do so would likely [...] increase complexity of the code by making it cache every file it comes across". It's solvable, but requires some meddling with selene, as it relies a lot on its current assumption of linting a file being an enclosed operation.

For example, having a cache of every file analyzed and using that for requires would make the multithreading code more complicated. Currently, selene just throws everything into thread pools, and it's extremely fast. If they all had to access a single cache, it'd be much slower.

**Comment by SvenMarcus** [comment]
Hm maybe a two step operation would be a good way to approach that.
Use a `HashSet` per thread to store the globals, merge the sets after all files have been analyzed and then adjust the errors if they turned out to be false positives.

**Comment by Kampfkarren** [comment]
That's better I think, but it still requires that the actual emissions be deferred all the way to the end, which is not ideal. 

I'm on the fence about this, but the most obvious solution to me right now is assuming idiomacy and making something like this optional as to not slow down execution on other codebases. I'm wary about this though, because it feels like it flies contrary to selene's "never be wrong" policy, and I'm uncomfortable with specific configurations like that.

**Comment by Kampfkarren** [comment]
One other thing--that solution won't be compatible with the VS Code extension, which analyzes individual files for speed (since it reinterprets the whole files every time).
